src/ir_passes/lower/compile.rs

```diff
@@ -0,0 +1,300 @@
+            }
+        }
+
+        // if this component is external, don't add new interface ports, as we must keep the signature the same
+        if !comp.is_ext {
+            // add remaining interface ports if not found (found ports already removed above)
+            for (attr, (name, width, dir)) in interface_ports {
+                ports.push(calyx::PortDef {
+                    name: (*name).into(),
+                    width: width_from_u64(*width),
+                    direction: dir.clone(),
+                    attributes: vec![*attr].try_into().unwrap(),
+                });
+            }
+        }
+
+        ports
+    }
+
+    /// Compiles a primitive component into a [calyx::Primitive]
+    fn primitive(ctx: &ir::Context, idx: ir::CompIdx) -> calyx::Primitive {
+        let comp = ctx.get(idx);
+
+        assert!(
+            comp.is_ext,
+            "Attempting to compile non-primitive component as primitive."
+        );
+
+        calyx::Primitive {
+            name: idx.name(ctx).into(),
+            params: comp
+                .params()
+                .iter()
+                .filter(|(_, p)| ir::ParamOwner::Sig == p.owner)
+                .map(|(idx, _)| idx.name(comp).into())
+                .collect(),
+            signature: Compile::ports(
+                ctx,
+                comp,
+                |value| calyx::Width::Const { value },
+                |comp, expr| expr.as_width(comp),
+            ),
+            attributes: calyx::Attributes::default(),
+            is_comb: false,
+            body: None,
+            latency: None,
+        }
+    }
+
+    /// Compiles an [ir::Component] into a [calyx::Component]
+    fn component(
+        ctx: &ir::Context,
+        idx: ir::CompIdx,
+        bind: &mut Binding,
+        lib: &calyx::LibrarySignatures,
+    ) -> calyx::Component {
+        log::debug!("Compiling component {idx}");
+        let comp = ctx.get(idx);
+
+        assert!(
+            !comp.is_ext,
+            "Attempting to compile primitive component as non-primitive."
+        );
+
+        let ports =
+            Compile::ports(ctx, comp, identity, |comp, expr| expr.as_u64(comp));
+        let mut component =
+            calyx::Component::new(idx.name(ctx), ports, false, false, None);
+        component.attributes.insert(calyx::BoolAttr::NoInterface, 1);
+
+        // If this is the main component, give it a `@top_level` attribute
+        if Some(idx) == ctx.entrypoint {
+            log::debug!("Defining main component {idx}");
+            component.attributes.insert(calyx::BoolAttr::TopLevel, 1);
+        }
+
+        let builder = calyx::Builder::new(&mut component, lib).not_generated();
+        let mut ctx = BuildCtx::new(ctx, idx, bind, builder, lib);
+
+        // Calculate the max states needed for each FSM for every event.
+        // Done by finding the furthest offset referenced in any [Time] in the component.
+        let mut max_states = HashMap::new();
```

Comment:
The max states computation needs to be moved out somewhere (maybe as a method on `Component`) because it is also used to generate `INTERFACE.json` when we're running our test harness. Currently everything works out because `fud/filament.py` doesn't use `--ir` with `--dump-interface` but it will break when we remove the ast-based compilation.

We can leave it here for now and extract in a following PR but please open an issue.